File: hist_equalizer_cpp_wo_opencv/hist_equalizer.cpp
#include "opencv2/core.hpp"
#include "opencv2/imgproc.hpp"
#include "opencv2/highgui.hpp"
#include "opencv2/video.hpp"
#include <iostream>
#include <stdio.h>
#include <chrono>
#include <cmath>

using namespace cv;
using namespace std::chrono;

static void calcHist(Mat& src, uint32_t histSize, double* histArray)
{
    for (int i = 0; i < histSize; i++)
        histArray[i] = 0.f;
    for (int y = 0; y < src.rows; y++)
        for (int x = 0; x < src.cols; x++)
        {
            int value = (int)src.at<uchar>(y, x);
            histArray[value]++;
        }
}

static void normalizeHist(uint32_t histSize, double* histArray, double newMax)
{
    double histMax = 0.f;
    for (int i = 0; i < histSize; i++)
        if (histMax < histArray[i])
            histMax = histArray[i];
    for (int i = 0; i < histSize; i++)
        histArray[i] *= (newMax/histMax);
}

static void makeCDF(uint32_t histSize, double* histArray, uint32_t cdfSize, double* cdfArray)
{
    for (int i = 0; i < cdfSize; i++)
        cdfArray[i] = 0;
    cdfArray[0] = histArray[0];
    for (int i = 1; i < cdfSize; i++)
        cdfArray[i] = cdfArray[i-1] + histArray[i];
}

static void equalizeHist(Mat& source, Mat& dst)
{
    double histArray[256];
    calcHist(source, 256, histArray);
    double cdfArray[256];
    makeCDF(256, histArray, 256, cdfArray);
    double eMap[256];
    double cdf_min = 0.f;
    int rows = source.rows, cols = source.cols;
    for (int i = 0; i < 256; i++)
        if (cdfArray[i] != 0)
        {
            cdf_min = cdfArray[i];
            break;
        }
    double k = 255.0f / (double)(rows*cols - cdf_min);
    for (int i = 0; i < 256; i++)
        eMap[i] = (k * (cdfArray[i] - cdf_min));
    for (int y = 0; y < rows; y++)
        for (int x = 0; x < cols; x++)
        {
            int value = (int)eMap[(int)source.at<uchar>(y, x)];
            // uchar value = 255;
            dst.at<uchar>(y, x) = value;
        }
}

static void make_hist_image(Mat& src, Mat& histImage)
{
    double histArray[256];
    calcHist(src, 256, histArray);
    int hist_w = histImage.cols;
    int hist_h = histImage.rows;
    normalizeHist(256, histArray, (double)hist_h);
    int bin_w = hist_w/256;
    for (int i = 1; i < 256; i++)
        line(histImage,
            Point2i(bin_w*(i-1), hist_h - (int)histArray[i-1]),
            Point2i(bin_w*i, hist_h - (int)histArray[i]),
            Scalar(255, 0, 0)
        );
}

int main(int argc, char const *argv[])
{
    long int frame_counter_equalized = 0;
    long int overall_time_equalized = 0;
    bool not_terminated = true;
    bool equalized = false;
    while(not_terminated)
    {
        Mat frame;
        VideoCapture cap;
        cap.open("/home/sailor/itmo_labs/cv/lab_1/samples/1.mp4");
        double fps = cap.get(CAP_PROP_FPS);
        double delay = 1000.f / fps;
        while (cap.isOpened())
        {
            auto timeNow = system_clock::now();
            auto startTime = duration_cast<milliseconds>(timeNow.time_since_epoch()).count();
            cap.read(frame);
            if (frame.empty()) {
                std::cerr << "ERROR! blank frame grabbed\n";
                break;
            }
            Mat gray;
            cvtColor(frame, gray, COLOR_BGR2GRAY);
            Mat histImg(512, 1024, CV_8UC1, Scalar(0, 0, 0));
            Mat histImg1(512, 1024, CV_8UC1, Scalar(0, 0, 0));
            Mat outImg = gray.clone();

            if (equalized)
            {
                timeNow = system_clock::now();
                auto start_equalization_time = duration_cast<milliseconds>(timeNow.time_since_epoch()).count();
                equalizeHist(gray, outImg);
                timeNow = system_clock::now();
                auto end_equalization_time = duration_cast<milliseconds>(timeNow.time_since_epoch()).count();
                int equalization_time = end_equalization_time - start_equalization_time;
                overall_time_equalized += equalization_time;
                frame_counter_equalized ++;
            }

            make_hist_image(outImg, histImg);
            imshow("hist", histImg);
            imshow("frame", outImg);
            timeNow = system_clock::now();
            auto endTime = duration_cast<milliseconds>(timeNow.time_since_epoch()).count();
            int execTime = endTime - startTime;
            std::cout << "Processing time of one frame: " << execTime << "ms" << std::endl;
            int timeout = max(1, (int)(delay - execTime));
            int k = waitKey(timeout);
            if (k == 113) // 'q'
            {
                std::cout << "Mean processing time of one equalized frame: " << (double)overall_time_equalized/(double)frame_counter_equalized << "ms" << std::endl;
                not_terminated = false;
                break;
            }
            if (k == 101)
            {
                equalized = !equalized;
                std::cout << "Equalized: " << equalized << std::endl;
            }
        }
    }
    return 0;
}

File: hist_equalizer_py_wo_opencv/hist_equalizer.py
import cv2 as cv
import numpy as np
import time

def current_milli_time():
    return round(time.time() * 1000)

def calcHist(source):
    hist = [0 for i in range(256)]
    rows, cols = source.shape
    for x in range(rows):
        for y in range(cols):
            value = source[x, y]
            hist[value] += 1
    return hist


def make_hist_image(src, hist_w, hist_h, hist_size):
    histRange = (0, 256) # the upper boundary is exclusive
    hist = cv.calcHist(src, [0], None, [hist_size], histRange)
    bin_w = int(round( hist_w/hist_size ))
    histImage = np.zeros((hist_h, hist_w, 3), dtype=np.uint8)
    cv.normalize(hist, hist, alpha=0, beta=hist_h, norm_type=cv.NORM_MINMAX)
    for i in range(1, hist_size):
        cv.line(histImage, ( bin_w*(i-1), hist_h - int(hist[i-1]) ),
            ( bin_w*(i), hist_h - int(hist[i]) ),
            ( 255, 0, 0), thickness=1)
    return histImage

def makeCDF(hist):
    cdf = [0 for i in range(len(hist))]
    cdf[0] = hist[0]
    for i in range(1, 256):
        cdf[i] = cdf[i-1] + hist[i]
    return cdf

def equalizeHist(source):
    hist = calcHist(source)
    rows, cols = source.shape
    cdf = makeCDF(hist)
    size = rows, cols, 1
    eMap = [0 for i in range(256)]
    cdf_min = 0
    for i in range(256):
        if (cdf[i] != 0):
            cdf_min = cdf[i]
            break
    k = 255.0 / (rows*cols - cdf_min)
    for i in range(256):
        eMap[i] = (k*(cdf[i] - cdf_min))
    equalizedImg = np.zeros(size, dtype=np.uint8)
    for x in range(rows):
        for y in range(cols):
            equalizedImg[x, y] = eMap[source[x, y]]
    return equalizedImg


not_terminated = True
overallTime = 0
frameCounter = 0
while (not_terminated):
    cap = cv.VideoCapture('/home/sailor/itmo_labs/cv/lab_1/samples/1.mp4')
    fps = cap.get(cv.CAP_PROP_FPS)
    delay = 1000 / fps
    equalized = False
    while cap.isOpened():
        startTime = current_milli_time()
        ret, frame = cap.read()
        if not ret:
            print("Can't receive frame (stream end?). Exiting ...")
            break
        src = cv.cvtColor(frame, cv.COLOR_BGR2GRAY)
        if equalized:
            startEqualizationTime = current_milli_time()
            src = equalizeHist(src)
            endEqualizationTime = current_milli_time()
            equalizationTime = endEqualizationTime - startEqualizationTime
            overallTime += equalizationTime
            frameCounter += 1
        srcHistImage = make_hist_image(src, 1024, 400, 256)
        cv.imshow('Source image', src)
        cv.imshow("Source Hist", srcHistImage)
        execTime = current_milli_time() - startTime
        print("Processing time of one frame: ", execTime , "ms")
        timeout = max(1, round(delay - execTime))
        k = cv.waitKey(timeout)
        if k == ord('q'):
            not_terminated = False
            print("Mean equalization time: ", overallTime / frameCounter, "ms")
            break
        if k == ord('e'):
            equalized = not equalized
            print("Equalized: ", equalized)

    cap.release()
cv.destroyAllWindows()
